### ISSUE TITLE
How to deal with multiple predictions (with different classes) on a same target

### ISSUE THREAD

**Issue by netw0rkf10w** [fizyr/keras-retinanet#466]
Hello,

First of all thank you so much for the very valuable implementation! It works consistently well on diverse problems for me.

I would like to ask for your advice on an issue that I encountered often when using the code: a target can get two predicted boxes, with two different classes. How should I best deal with this?

Of course the simplest solution is to keep only the one with the highest score. I'm also thinking of another solution that is to take the average of the classification output vectors, then use that vector to predict the class.

Could I have your opinion please?
Thank you again!

**Comment by hgaiser** [comment]
> I'm also thinking of another solution that is to take the average of the classification output vectors, then use that vector to predict the class.

I'm not sure I follow this approach? The classification vector is what is being used to perform the classification. If more than one of those classes have high scores, they are indeed two detection. This is by design actually, we didn't use to have that before (we'd only accept the maximum score for a classification vector, now we take any score above a threshold). I think the idea is that two objects can have very similar bounding boxes but still be separate classes.

I would guess that if your network is outputting two different classes for the same object, your objects are visually similar? Or maybe there are some mislabelled object in your dataset? Or maybe you don't have enough data for your use case?

There are many possibilities :)

**Comment by netw0rkf10w** [comment]
@hgaiser Thanks for the answer!

> I'm not sure I follow this approach? The classification vector is what is being used to perform the classification. If more than one of those classes have high scores, they are indeed two detection. This is by design actually, we didn't use to have that before (we'd only accept the maximum score for a classification vector, now we take any score above a threshold).

If I understand correctly, you mean each bounding box can be assigned to two (or more) different classes? The problem I described above is a bit different though: I have two slightly different boxes (in terms of position), each is assigned to a different classes. Thus, I guess that these two boxes have two different classification vectors (because the boxes are not identical), so I can take the average of these vectors, hoping that will yield the true class (for my task I would like to have only one detection per object).

> I would guess that if your network is outputting two different classes for the same object, your objects are visually similar? Or maybe there are some mislabelled object in your dataset? Or maybe you don't have enough data for your use case?

Well all of these are likely, unfortunately!

My datset probably contains some noise, but I would say the level is not high. A more serious problem is the lack of diversity and balance: (a) the dataset contains only images of single class (i.e. each image contains only one class of objects, while the model will be used on images of different classes), and (b) the number of images can be very different between classes.

I don't know how much (a) may affect the detection accuracy, but I have a solution in mind that is to create synthetic images of mixtures of different classes.

For (b) what I did is to create duplicates of images of minority classes so that the numbers of images of all classes are roughly the same. One question please: should I better have the same **number of images** for each class, or rather the same **number of ground-truth bounding boxes**?
Instead of creating duplicates, I think it's better to do 'weighted sampling' when creating training batches (e.g. for classification in Keras: `model.fit(X_train, y_train, class_weight=class_weight)`), but I couldn't find such option from the code :(

Thank you so much again for your help!

**Comment by netw0rkf10w** [comment]
@hgaiser Thank you so much for your answer!

**Comment by LaurensHagendoorn** [comment]
I am hijackign this because it seems similar. I am getting this issue only since I last updated (been a while).
![image](https://user-images.githubusercontent.com/32707472/41167131-55c634cc-6b42-11e8-8433-6d9638ed618d.png)
Here it predicted that it is both a 4 and a 6, whereas this never used to happen before, and I would think that this is also incorrect behaviour? If this is not a bug then you can ignore it (I'll fix it myself) but I've never encountered this in the old version I was using.

**Comment by hgaiser** [comment]
Hmm how old is the older version you were using? We changed NMS to not compete amongst classes a while back. That might explain the difference. We're working on making this change optional actually.

**Comment by netw0rkf10w** [comment]
I confirm that this problem did not occur in December. But as @hgaiser said, NMS has been changed so that's probably the reason.

**Comment by LaurensHagendoorn** [comment]
haha very old, I have changed so much by now that it is quite the hastle to update ;)
It sounds like that is the change, so I'll disable that in my code.
thanks!

**Comment by netw0rkf10w** [comment]
@hgaiser Sorry it looks like the problem does not lie in the code but in my dataset (which is a very tiny one that I used for debugging). I'm trying again on a large dataset to see.

**Comment by hgaiser** [comment]
I think this will solve your issue: https://github.com/fizyr/keras-retinanet/pull/512

**Comment by netw0rkf10w** [comment]
@hgaiser That is awesome! Thanks a lot! :+1:

**Comment by hgaiser** [comment]
Feedback is welcome, I haven't had a chance to test it yet ^^

**Comment by netw0rkf10w** [comment]
@hgaiser Sure. I'm going to try it now and will get back in a few hours. Thanks again!

**Comment by netw0rkf10w** [comment]
@hgaiser It seems to me that `train.py` hasn't been updated accordingly?

**Comment by hgaiser** [comment]
You mean the mAP computation in `train.py`? It's not as flexible as `convert_model.py`, that's true. I'm looking into better methods for configuration, so I'm a bit hesitant to add every flag to every tool. Could you test it by using `convert_model.py` and then running `evaluate.py` ?

**Comment by netw0rkf10w** [comment]
@hgaiser Sorry it took me some time to check.

I have converted a model that was trained with the previous code version (i.e. before your pull request) using the new code:

`keras_retinanet/bin/convert_model.py ./snapshots/resnet101_csv_22.h5 ./inference_models/resnet101_csv_inference_22.h5 --backbone resnet101 --no-class-specific-filter`

and evaluated it:

`keras_retinanet/bin/evaluate.py --backbone resnet101 --max-detections 200  csv val_annotations.csv names.csv ./inference_models/resnet101_csv_inference_22.h5`

without any errors. In addition, I got higher mAP (which is expected because my images contain contain only objects without big overlaps).

**However**, when testing the model on some external images:

```
model = models.load_model(model_path, backbone_name=backbone, class_specific_filter=False)
...
boxes, scores, labels = model.predict_on_batch(np.expand_dims(image, axis=0))
...
```

I still got duplicated boxes. I guess this is because the `non_max_suppression` function works independently for each class? Maybe I'll have to remove duplicates manually as a post-processing step.

**Comment by netw0rkf10w** [comment]
> You mean the mAP computation in train.py?

I meant for consistency some function calls in `train.py` can be updated as well, e.g. line 112:

```
# make prediction model
prediction_model = retinanet_bbox(model=model)
```

But if the modification of FilterDetections does not affect the training (except for the mAP computation) then it's not a major issue.

**Comment by netw0rkf10w** [comment]
There is also something very strange:

Using the old code, I tried modifying `filter_detections.py` to let the class compete, like I said in a [previous comment](https://github.com/fizyr/keras-retinanet/issues/466#issuecomment-398113094). I based on your code (which is different than mine at the before-last line):

```
    scores  = keras.backend.max(classification, axis    = 1)
    labels  = keras.backend.argmax(classification, axis = 1)
    # threshold based on score
    indices = backend.where(keras.backend.greater(scores, score_threshold))

    if nms:
        filtered_boxes  = backend.gather_nd(boxes, indices)
        filtered_scores = keras.backend.gather(scores, indices)[:, 0]

        # perform NMS
        nms_indices = backend.non_max_suppression(filtered_boxes, filtered_scores, max_output_size=max_detections, iou_threshold=nms_threshold)

        # filter indices based on NMS
        indices = keras.backend.gather(indices, nms_indices)

    # add indices to list of all indices
    labels = backend.gather_nd(labels, indices)
    indices = keras.backend.stack([indices[:, 0], labels], axis=1)
```
This is supposed to give identical results to your new code with `class_specific_filter=False`. However, when converting a model and testing it using the modified code on some images, I obtained very few boxes (with no duplicates). With your new code, I obtained all the (good) boxes but with duplicates.

**Comment by netw0rkf10w** [comment]
I have checked again and realized that, with or without `--no-class-specific-filter` (for `convert_model.py`), the results are the same :( I haven't found the cause yet.

**Comment by hgaiser** [comment]
Could you share an example of these slightly different positions? NMS works with an overlap of 0.5, would you say it is within 0.5?

**Comment by netw0rkf10w** [comment]
@hgaiser I forgot to mention that the boxes are of different classes. NMS of one class cannot suppress another class I guess.

![retina](https://user-images.githubusercontent.com/8569515/41720813-81f3cfce-7564-11e8-9263-e7b6f7873168.jpg)

**Comment by hgaiser** [comment]
Those should definitely be suppressed when `class_specific_filter=False`. This is the result using #512 ? Could you double check that the value is correctly set inside the [`FilterDetections`](https://github.com/fizyr/keras-retinanet/pull/512/files#diff-aa9bbc37b19ffadced9c6515791d4049R71) layer?

Also, maybe we should continue the discussion at #512 :)

**Comment by netw0rkf10w** [comment]
Thanks. I've posted a comment in #512. Let's continue there ;)
